docs/advanced/signal_generators.md

```diff
@@ -0,0 +1,558 @@
+
+Arbitrary Function Generators (AFGs) utilize a phase increment process and a data lookup table to provide variable frequencies.
+The phase increment calculated is dependent on the waveform length and frequency requested. This has
+a side effect where the phase increment can be larger than one index in the lookup table. Functions bypass this
+issue by being simplistic enough that waveform length reduction doesn't have a detrimental effect on the end output.
+
+Arbitrary Waveform Generators (AWGs) enforce one cycle per sample, allowing the output to be the same shape regardless of clock rate.
+With low frequency functions, AWGs are functionally identical to AFGs, besides offering more constrained amplitudes and offsets.
+
+______________________________________________________________________
+
+### Class Structure
+
+```{mermaid}
+classDiagram
+    direction LR
+
+    SignalGeneratorMixin <|-- Tekscope
+    SignalGeneratorMixin <|-- AFG
+    SignalGeneratorMixin <|-- AWG
+
+```
+
+The `SignalGenerator` class is responsible for most waveform generators, including the `AWG` and `AFG`.
+Similarly, `TekScope` is responsible for the AFG's internal to the scopes themselves, commonly referred to as
+an IAFG. All of these classes inherit `SignalGeneratorMixin`, which includes a list of methods which share
+functionality throughout all signal sources.
+
+```{note}
+SignalGeneratorMixin only contains abstract methods, defining the class by itself and calling methods in it will only raise NotImplemented errors.
+```
+
+```{mermaid}
+classDiagram
+    direction LR
+
+    Tekscope <|-- InternalAFGChannel
+    AFG <|-- AFGSourceChannel
+
+    AWG <|-- AWGSourceChannel
+    AWGSourceChannel <|-- AWG5KSourceChannel
+    AWG5KSourceChannel <|-- AWG7KSourceChannel
+    AWGSourceChannel <|-- AWG5200SourceChannel
+    AWGSourceChannel <|-- AWG70KSourceChannel
+```
+
+Each Source class (`AFG`, `AWG`) and `Tekscope` (If the AFG license is installed) will contain a dictionary of source channel classes,
+which are defined on first access. Each of these source channel classes represents an output on the source, or the IAFG in the case
+of an oscilloscope.
+
+These channels contain methods and properties which pertain to PI commands which only apply changes to one output channel.
+For example: the afg.source_channel\["SOURCE1"\].set_amplitude() call will change the amplitude only for source output 1.
+
+```{tip}
+These channel classes not only provide easy access to basic scipy commands, but also helper functions, like `set_function_properties`
+```
+
+______________________________________________________________________
+
+### Class Methods
+
+```{warning}
+Each method exclude most attempts at validation, as the end user can change aspects outside its purview.
+There are several distinct instances where this can cause unwanted behavior, depending on the source and what
```

Comment:
```suggestion
There are several distinct instances where this can cause unwanted behavior, depending on the signal generator and what
```

There may be other instances of source that need to be changed to signal generator.